presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/LogFilter.java

```diff
@@ -0,0 +1,286 @@
+        }
+
+        Stream<StringJoiner> streamEntries()
+        {
+            return entries.stream();
+        }
+
+        /**
+         * Add a line from a log file to a deque of log entries
+         *
+         * If the string does not match {@link #logPattern}, it will be
+         * appended to the previous entry. If there is no previous entry, a new
+         * entry will be created only if entries are not being filtered.
+         *
+         * If {@link #maxEntries} is non-null and the deque contains that many
+         * entries, the oldest will be removed before the new entry is added.
+         */
+        void addLine(String line)
+        {
+            Matcher matcher = logPattern.matcher(line);
+            if (matcher.matches()) {
+                /* Apply all filters */
+                boolean passing = true;
+                for (ImmutableMap.Entry<String, Predicate<String>> f
+                        : namedGroupFilters.entrySet()) {
+                    passing = passing && f.getValue().test(matcher.group(f.getKey()));
+                }
+                for (ImmutableMap.Entry<Integer, Predicate<String>> f
+                        : numberedGroupFilters.entrySet()) {
+                    passing = passing && f.getValue().test(matcher.group(f.getKey()));
+                }
+
+                if (passing) {
+                    keptLastLine = addNewEntry(line);
+                }
+                else {
+                    keptLastLine = false;
+                }
+            }
+            else if (keptLastLine == null
+                    && namedGroupFilters.isEmpty() && numberedGroupFilters.isEmpty()) {
+                // If the first line does not match the pattern and there are no filters,
+                // use the line as a new entry.
+                keptLastLine = addNewEntry(line);
+            }
+            else if (keptLastLine != null && keptLastLine) {
+                keptLastLine = addToLast(line);
+            }
+            else {
+                keptLastLine = false;
+            }
+        }
+
+        /**
+         * @return true if the line addition succeeded
+         */
+        private boolean addNewEntry(String line)
+        {
+            if (maxEntries != null && entries.size() >= maxEntries) {
+                if (keepFirst) {
+                    return false;
+                }
+                else if (maxEntries == 0) {
+                    return true;
+                }
+                entries.removeFirst();
+            }
+            StringJoiner joiner = new StringJoiner(lineSeparator);
+            joiner.add(line);
+            entries.addLast(joiner);
+            return true;
+        }
+
+        /**
+         * @return true if the line addition succeeded
+         * (equivalently, true if any entries already exist)
+         */
+        private boolean addToLast(String line)
+        {
+            boolean isNotEmpty = !entries.isEmpty();
```

Comment:
It seems a bit more readable to me this way. (Though I've just renamed the variable to `entriesExist`, which makes it a bit clearer.)